### ISSUE TITLE
Cop Idea: fetch format

### ISSUE THREAD

**Issue by mockdeep** [rubocop-hq/rubocop#3353]
When using `Hash#fetch`, there is a block format and an argument format. It would be nice to push the argument format when the default is a constant:

``` ruby
# good
hash.fetch(:foo, 52)

# bad
hash.fetch(:foo) { 52 }
```

**Comment by bbatsov** [comment]
That's not a good idea. If you have to calculate something expensive as the default value you'll still need the block form to delay that computation. For something like this to make sense you'll have to be certain the default value is some literal or maybe a constant. I also envision we might run into issues with other methods being named `fetch`.

**Comment by mockdeep** [comment]
Yeah, I wouldn't want to do it for non-constant values, but can't you tell if it's `nil` or `true` or `52`? Dealing with other `fetch` method implementations seems a little trickier, though. Is there no way to tell if an object is a hash? It's probably much more likely to be, though, if the block syntax is used...

**Comment by mikegee** [comment]
> Is there no way to tell if an object is a hash? It's probably much more likely to be, though, if the block syntax is used... 

Yeah, it would be hard to distinguish `Hash#fetch` from [`ActiveSupport::Cache::Store#fetch`](http://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch). They can both take one argument and a block.

**Comment by Drenmi** [comment]
> Yeah, it would be hard to distinguish Hash#fetch from ActiveSupport::Cache::Store#fetch. They can both take one argument and a block.

We can always give it a `SafeMode` until `rubocop-rails` becomes reality. 😀

**Comment by mockdeep** [comment]
It looks like `ActiveSupport::Cache::Store` has the same mechanics, though, so the rule might apply there as well.

**Comment by bbatsov** [comment]
I'm closing this ticket due to no recent activity. Feel free to re-open it if you ever come back to it.

**Comment by mockdeep** [comment]
Still interested in this rule.

**Comment by bbatsov** [comment]
PRs welcome! 😉
